src/XAD/Hessian.hpp

```diff
@@ -0,0 +1,202 @@
+/*******************************************************************************
+
+   Implementation of Hessian computing methods.
+
+   This file is part of XAD, a comprehensive C++ library for
+   automatic differentiation.
+
+   Copyright (C) 2010-2024 Xcelerit Computing Ltd.
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU Affero General Public License as published
+   by the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+******************************************************************************/
+
+#include <functional>
+#include <vector>
+
+namespace xad
+{
+
+template <class T>
+class Hessian
+{
+  public:
+    // fwd_adj 2d vector constructor
+    Hessian(std::function<T(std::vector<T> &)> func, const std::vector<T> &v,
+            xad::Tape<xad::FReal<double>> *tape)
```

Comment:
Can you explain what the type `T` must be here? Is that an XAD type, or double/float? Does it have to be an `xad::FReal<...>` type? Or the second-order type? I think that latter, right?

If so, you can:
- use [traits](https://github.com/auto-differentiation/xad/blob/main/src/XAD/Literals.hpp#L147) and [type members](https://github.com/auto-differentiation/xad/blob/main/src/XAD/Literals.hpp#L169) to deduce the tape type from that. 
- use [getActive()](https://github.com/auto-differentiation/xad/blob/main/src/XAD/Tape.hpp#L102) to get the currently active instance. 
- You might even support the case where no tape has been setup externally (`getActive()` returns `nullptr`), and set one up inside compute instead.
- from those traits, you'd see if it's `fwd_fwd` or `fwd_adj` automatically

Reply:
You're right `T` is the second-order type, so AReal<FReal<double>>. Could you give me some more direction on how to do what you described?

> use [traits](https://github.com/auto-differentiation/xad/blob/main/src/XAD/Literals.hpp#L147) and [type members](https://github.com/auto-differentiation/xad/blob/main/src/XAD/Literals.hpp#L169) to deduce the tape type from that.

From my understanding the goal is to reduce the use of `T` by deducing types. I'm just not sure how to extract them using the traits.

would that mean the constructor becomes something like:
```c++
    // fwd_adj 2d vector constructor
    Hessian(std::function<T(std::vector<T> &)> func, const std::vector<T> &v,
            xad::Tape</*some deduction method*/> *tape = Tape::getActive())
        : foo_(func),
          v_(v),
          domain_(static_cast<unsigned int>(v_.size())),
          matrix_(domain_, std::vector<T>(domain_, 0.0))
    {
        compute(tape);
    }
```

and then `compute()` overloads are resolved by either `AD` or `tape_type`? Again, I'm not sure how to make these comparisons and what I've tried so far doesn't work.

Reply:
I think if you make the code functional, it will be easier and you don't even need this type deduction. You can just use overloads, something along those lines:

```c++
vector<vector<T>> computeHessian(
   std::function<AReal<FReal<T>>(const std::vector<AReal<FReal<T>>&)> func, 
   const std::vector<AReal<FReal<T>>& v,
   typename AReal<FReal<T>>::tape_type* tape = typename AReal<FReal<T>>::tape_type::getActive()
);
```

Basically have `T` point to the underlying scalar. And do another for `fwd_fwd`. Note that the return type would be the underlying scalar type, not the active type (which makes sense).

Reply:
Understood. I've made a commit with changes applying this advice.

Note the return type in your snippet is `vector<vector<T>>`. Is this intended or did you mean `vector<vector<AReal<FReal<T>>>>`?

Also, now that I have no longer the class as a boilerplate, making `v` constant prevents me from accessing the derivatives. I could make a copy of `v` within the method (which was essentially what was happening in the previous implementation) and that would fix the issue.

Reply:
You also mentioned deducing tape type, supporting no tape scenarios and deducing `fwd_fwd` or `fwd_adj` from that. I've thought about it and is that not inefficient? Setting up a tape/getActive() within the method:
- causes larger overhead
- can lead to ambiguous error prone code (when the active tape type doesn't correspond any of the methods)
- possibly allocation of an additional thread for a new a tape when a preexisting one would suffice
- increases auxiliary space and function calls since the overload is resolved within an intermediary
 
 I can explain in doc the proper usage if you agree.

Reply:
We believe the return type `vector<vector<T>>` is correct. The term `derivative(value(x))` returns type `T`. The active type is only for operator overloading to compute things internally, but the user will be interested in the Hessian/Jacobian in `double`. 

Also making a copy of the input vector is ok in this case. It has a small overhead, but compared to compute the Hessian of a complex function, that will be barely measurable. We could in fact even allow users to pass in the inputs a `vector<double>` (matching the output), it wouldn't make much of a difference. But not important at this stage.

Regarding the tape initialisation within the function - it is optional. If users really only want the Hessian at one input point, then it makes no difference whether the tape is initialised inside our outside of the function. If they do more with the tape, then they have to option to initialise it themselves and pass it in. You could do it along these lines:

```c++
... computeHessian(..., tape_type* tape = tape_type::getActive()) 
{
  std::unique_ptr<tape_type> t;
  if (tape == nullptr)
  {
     t = std::make_unique<tape_type>();  // constructor activates the tape automatically
     tape = t.get();
  }
  ...
}   // unique_ptr destructs automatically if not nullptr
```